README.md
# &#x1F4E6; Biip

_Biip interprets the data in barcodes._

[![Tests](https://img.shields.io/github/workflow/status/jodal/biip/Tests)](https://github.com/jodal/biip/actions?workflow=Tests)
[![Docs](https://img.shields.io/readthedocs/biip)](https://biip.readthedocs.io/en/latest/)
[![Coverage](https://img.shields.io/codecov/c/gh/jodal/biip)](https://codecov.io/gh/jodal/biip)
[![PyPI](https://img.shields.io/pypi/v/biip)](https://pypi.org/project/biip/)

---

Biip is a Python library for making sense of the data in barcodes.

The library can interpret the following formats:

- GTIN-8, GTIN-12, GTIN-13, and GTIN-14 numbers,
  commonly found in EAN-8, EAN-13, UPC-A, UPC-E, and ITF-14 barcodes.

- GS1 AI element strings,
  commonly found in GS1-128 barcodes.

For a quickstart guide and a complete API reference,
see the [documentation](https://biip.readthedocs.io/).

## Installation

Biip requires Python 3.6 or newer.

Biip is available from [PyPI](https://pypi.org/project/biip/):

```
python3 -m pip install biip
```

Optionally, with the help of `py-moneyed`, Biip can convert amounts with
currency information to `moneyed.Money` objects.
To install Biip with `py-moneyed`, run:

```
python3 -m pip install "biip[money]"
```

# Project resources

- [Documentation](https://biip.readthedocs.io/)
- [Source code](https://github.com/jodal/biip)
- [Releases](https://github.com/jodal/biip/releases)
- [Issue tracker](https://github.com/jodal/biip/issues)
- [Contributors](https://github.com/jodal/biip/graphs/contributors)
- [Users](https://github.com/jodal/biip/wiki/Users)

## Development status

All planned features have been implemented.
Please open an issue if you have and barcode parsing related needs that are not covered.

## Features

- GS1 (multiple Element Strings with Application Identifiers)
  - Recognize all specified Application Identifiers.
  - Recognize allocating GS1 Member Organization from the GS1 Company Prefix.
  - Parse fixed-length Element Strings.
  - Parse variable-length Element Strings.
    - Support configuring the separation character.
  - Parse AI `00` as SSCC.
  - Parse AI `01` and `02` as GTIN.
  - Parse dates into `datetime.date` values.
    - Interpret the year to be within -49/+50 years from today.
    - Interpret dates with day "00" as the last day of the month.
  - Parse variable measurement fields into `Decimal` values.
  - Parse discount percentage into `Decimal` values.
  - Parse amounts into `Decimal` values.
    - Additionally, if py-moneyed is installed,
      parse amounts with currency into `Money` values.
  - Encode as Human Readable Interpretation (HRI),
    e.g. with parenthesis around the AI numbers.
  - Easy lookup of parsed Element Strings by:
    - Application Identifier (AI) prefix
    - Part of AI's data title
- GTIN (Global Trade Item Number)
  - Parse GTIN-8, e.g. from EAN-8 barcodes.
  - Parse GTIN-12, e.g. from UPC-A and UPC-E barcodes.
  - Parse GTIN-13, e.g. from EAN-13 barcodes.
  - Parse GTIN-14, e.g. from ITF-14 and GS1-128 barcodes.
  - Extract and validate check digit.
  - Extract GS1 Prefix.
  - Extract packaging level digit from GTIN-14.
  - Encode GTIN-8 as GTIN-12/13/14.
  - Encode GTIN-12 as GTIN-13/14.
  - Encode GTIN-13 as GTIN-14.
- RCN (Restricted Circulation Numbers), a subset of GTINs
  - Classification of RCN usage to either a geographical region or a company.
  - Parsing of variable measurements (price/weight) into `Decimal`
    values.
  - Parsing of price values into `Money` values if `py-moneyed` is
    installed and the region's RCN parsing rules specifies a currency.
  - Baltics: Parsing of weight.
  - Great Britain: Parsing of price, including validation of price check digit.
  - Norway: Parsing of weight and price.
  - Sweden: Parsing of weight and price.
  - Encode RCN with the variable measure part zeroed out,
    to help looking up the correct trade item.
- SSCC (Serial Shipping Container Code)
  - Validate check digit.
  - Encode for human consumption, with the logical groups separated by whitespace.
- Symbology Identifers, e.g. `]EO`
  - Recognize all specified Symbology Identifier code characters.
  - Strip Symbology Identifers before parsing the remainder.
  - Use Symbology Identifers when automatically selecting what parser to use.

## License

Biip is copyright 2020-2021 Stein Magnus Jodal and contributors.
Biip is licensed under the
[Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0).


docs/changes.rst
=======
Changes
=======

The changes between Biip versions can be found at the
`GitHub release page <https://github.com/jodal/biip/releases>`_.


src/biip/gtin.py
"""Support for Global Trade Item Number (GTIN).

The :mod:`biip.gtin` module contains biip's support for parsing GTIN formats.

A GTIN is a number that uniquely identifies a trade item.

This class can interpet the following GTIN formats:

- GTIN-8, found in EAN-8 barcodes.
- GTIN-12, found in UPC-A and UPC-E barcodes.
- GTIN-13, found in EAN-13 barcodes.
- GTIN-14, found in ITF-14 barcodes, as well as a data field in GS1 barcodes.

A GTIN can be converted to any other GTIN format, as long as the target
format is longer.

Example:
    >>> from biip.gtin import Gtin
    >>> gtin = Gtin.parse("5901234123457")
    >>> gtin
    Gtin(value='5901234123457', format=GtinFormat.GTIN_13,
    prefix=GS1Prefix(value='590', usage='GS1 Poland'),
    payload='590123412345', check_digit=7, packaging_level=None)
    >>> gtin.as_gtin_14()
    '05901234123457'
"""

from __future__ import annotations

from dataclasses import dataclass, field
from decimal import Decimal
from enum import Enum, IntEnum
from typing import Optional, Type, Union

from biip import EncodeError, ParseError
from biip.gs1 import GS1Prefix
from biip.gs1.checksums import numeric_check_digit

try:
    import moneyed
except ImportError:  # pragma: no cover
    moneyed = None


__all__ = ["Gtin", "GtinFormat", "Rcn", "RcnUsage"]


class GtinFormat(IntEnum):
    """Enum of GTIN formats."""

    #: GTIN-8
    GTIN_8 = 8

    #: GTIN-12
    GTIN_12 = 12

    #: GTIN-13
    GTIN_13 = 13

    #: GTIN-14
    GTIN_14 = 14

    def __str__(self: GtinFormat) -> str:
        """Pretty string representation of format."""
        return self.name.replace("_", "-")

    def __repr__(self: GtinFormat) -> str:
        """Canonical string representation of format."""
        return f"GtinFormat.{self.name}"

    @property
    def length(self: GtinFormat) -> int:
        """Length of a GTIN of the given format."""
        return int(self)


class RcnUsage(Enum):
    """Enum of RCN usage restrictions."""

    #: Usage of RCN restricted to geopgraphical area.
    GEOGRAPHICAL = "geo"

    #: Usage of RCN restricted to internally in a company.
    COMPANY = "company"


@dataclass
class Gtin:
    """Data class containing a GTIN."""

    #: Raw unprocessed value.
    #:
    #: May include leading zeros.
    value: str

    #: GTIN format, either GTIN-8, GTIN-12, GTIN-13, or GTIN-14.
    #:
    #: Classification is done after stripping leading zeros.
    format: GtinFormat

    #: The GS1 prefix, indicating what GS1 country organization that assigned
    #: code range.
    prefix: GS1Prefix

    #: The actual payload, including packaging level if any, company prefix,
    #: and item reference. Excludes the check digit.
    payload: str

    #: Check digit used to check if the GTIN as a whole is valid.
    check_digit: int

    #: Packaging level is the first digit in GTIN-14 codes.
    #:
    #: This digit is used for wholesale shipments, e.g. the GTIN-14 product
    #: identifier in GS1-128 barcodes, but not in the GTIN-13 barcodes used for
    #: retail products.
    packaging_level: Optional[int] = None

    @classmethod
    def parse(cls: Type[Gtin], value: str) -> Gtin:
        """Parse the given value into a :class:`Gtin` object.

        Both GTIN-8, GTIN-12, GTIN-13, and GTIN-14 are supported.

        Args:
            value: The value to parse.

        Returns:
            GTIN data structure with the successfully extracted data.
            The checksum is guaranteed to be valid if a GTIN object is returned.

        Raises:
            ParseError: If the parsing fails.
        """
        if len(value) not in (8, 12, 13, 14):
            raise ParseError(
                f"Failed parsing {value!r} as GTIN: "
                f"Expected 8, 12, 13, or 14 digits, got {len(value)}."
            )

        if not value.isnumeric():
            raise ParseError(
                f"Failed parsing {value!r} as GTIN: Expected a numerical value."
            )

        stripped_value = _strip_leading_zeros(value)
        assert len(stripped_value) in (8, 12, 13, 14)

        num_significant_digits = len(stripped_value)
        gtin_format = GtinFormat(num_significant_digits)

        payload = stripped_value[:-1]
        check_digit = int(stripped_value[-1])

        packaging_level: Optional[int] = None
        if gtin_format == GtinFormat.GTIN_14:
            packaging_level = int(stripped_value[0])
            value_without_packaging_level = stripped_value[1:]
            prefix = GS1Prefix.extract(value_without_packaging_level)
        elif gtin_format == GtinFormat.GTIN_8:
            prefix = GS1Prefix.extract(stripped_value.zfill(12))
        else:
            prefix = GS1Prefix.extract(stripped_value)

        calculated_check_digit = numeric_check_digit(payload)
        if check_digit != calculated_check_digit:
            raise ParseError(
                f"Invalid GTIN check digit for {value!r}: "
                f"Expected {calculated_check_digit!r}, got {check_digit!r}."
            )

        gtin_type: Type[Union[Gtin, Rcn]]
        if "Restricted Circulation Number" in prefix.usage:
            gtin_type = Rcn
        else:
            gtin_type = Gtin

        return gtin_type(
            value=value,
            format=gtin_format,
            prefix=prefix,
            payload=payload,
            check_digit=check_digit,
            packaging_level=packaging_level,
        )

    def as_gtin_8(self: Gtin) -> str:
        """Format as a GTIN-8."""
        return self._as_format(GtinFormat.GTIN_8)

    def as_gtin_12(self: Gtin) -> str:
        """Format as a GTIN-12."""
        return self._as_format(GtinFormat.GTIN_12)

    def as_gtin_13(self: Gtin) -> str:
        """Format as a GTIN-13."""
        return self._as_format(GtinFormat.GTIN_13)

    def as_gtin_14(self: Gtin) -> str:
        """Format as a GTIN-14."""
        return self._as_format(GtinFormat.GTIN_14)

    def _as_format(self: Gtin, gtin_format: GtinFormat) -> str:
        if self.format.length > gtin_format.length:
            raise EncodeError(
                f"Failed encoding {self.value!r} as {gtin_format!s}."
            )
        return f"{self.payload}{self.check_digit}".zfill(gtin_format.length)


@dataclass
class Rcn(Gtin):
    """Restricted Circulation Number (RCN) is a subset of GTIN.

    RCNs with prefix 02 and 20-29 have the same semantics across a geographic
    region, defined by the local GS1 Member Organization.

    RCNs with prefix 40-49 have semantics that are only defined within a
    single company.
    """

    #: Where the RCN can be circulated,
    #: in a geographical area or within a company.
    usage: Optional[RcnUsage] = field(default=None, init=False)

    #: A variable weight value extracted from the barcode,
    #: if indicated by prefix.
    weight: Optional[Decimal] = field(default=None, init=False)

    #: A variable weight price extracted from the barcode,
    #: if indicated by prefix.
    price: Optional[Decimal] = field(default=None, init=False)

    #: A Money value created from the variable weight price.
    #: Only set if py-moneyed is installed and the currency is known.
    money: Optional["moneyed.Money"] = field(default=None, init=False)

    def __post_init__(self: Rcn) -> None:
        """Initialize derivated fields."""
        self._set_usage()

    def _set_usage(self: Rcn) -> None:
        if "within a geographic region" in self.prefix.usage:
            self.usage = RcnUsage.GEOGRAPHICAL
        elif "within a company" in self.prefix.usage:
            self.usage = RcnUsage.COMPANY


def _strip_leading_zeros(value: str) -> str:
    if len(value) in (12, 13, 14) and len(value.lstrip("0")) in (9, 10, 11, 12):
        # Keep up to three leading zeros in GTIN-12
        num_zeros_before_gtin_12 = len(value) - 12
        return value[num_zeros_before_gtin_12:]

    if len(value) >= 8 and len(value.lstrip("0")) <= 8:
        # Keep all leading zeros in GTIN-8
        num_zeros_before_gtin_8 = len(value) - 8
        return value[num_zeros_before_gtin_8:]

    return value.lstrip("0")
